### DISCUSSION TITLE
How to align a graph in a more centered way

### DISCUSSION THREAD

**Discussion by pbrzosko** [erikbrinkman/d3-dag#117] [Q&A]
Hi,

I have a simple graph and it renders like that:

<img width="1071" alt="image" src="https://github.com/erikbrinkman/d3-dag/assets/67122459/f940a3df-6892-438f-a044-dd6ca67a767b">

How can I make it to render more like that, of course please don't bother about broken links, it is more about node placement:

![image](https://github.com/erikbrinkman/d3-dag/assets/67122459/fd7c2638-4c01-4fe5-b431-6679c1993278)

**Comment by pbrzosko** [comment]
```const builder = graphStratify();
  const graph = builder(data);
  const layout = sugiyama()
    // @ts-ignore
    .nodeSize([constants.BASE_HEIGHT, constants.BASE_WIDTH])
    .gap([50, 50])
    .layering(
      layeringSimplex().rank(
        (node: GraphNode<{ passage: number }>) => node.data.passage
      )
    )
    .tweaks([tweakFlip("diagonal")])
    .coord(coordCenter());```

**Reply by pbrzosko** [reply]
And also why do I get a TS error when I do `sugiyama().nodeSize()`. Looking at DefaultSugiyama type that is returned from sugiyama there is not nodeSize() function, is just readonly property `nodeSize: readonly [1, 1];`, so TS seems to be correct?

**Reply by erikbrinkman** [reply]
Can you post the exact error? It's harder to diagnose without seeing what the problem is.

**Reply by pbrzosko** [reply]
![image](https://github.com/erikbrinkman/d3-dag/assets/67122459/0273ea64-4ca9-4f6d-942b-f44d3a2be58c)

**Comment by erikbrinkman** [comment]
The layout looks like that because you're using `coordCenter()`, try using a different coordinate assignment algorithm, like the default, and you should get something closer to what you expect.

**Reply by pbrzosko** [reply]
Hi, the reason I started to playaround with coords was because the graph looked dxactly the same by default...

**Reply by erikbrinkman** [reply]
It should look significantly different, in [this example](https://observablehq.com/@erikbrinkman/d3-dag-sugiyama) you can see the stark difference between simplex (the default) and center (what you're using here)

**Reply by pbrzosko** [reply]
OK, so I think diagonal flip breaks it. Here is the result without tweaks and coord:
![image](https://github.com/erikbrinkman/d3-dag/assets/67122459/ccf7c596-9c18-4e71-9856-c215da3b46f4)
But still nodes are not centered in a way that two branches are equally aligned to the parent?

and when I apply a tweak, it is like that:
![image](https://github.com/erikbrinkman/d3-dag/assets/67122459/beb53268-b06c-43b5-ac07-ae635d1d8611)


So I assumed I needed to switch node width with height and this caused the issue. So I reverted it back to `.nodeSize([constants.BASE_WIDTH, constants.BASE_HEIGHT])` but it ends up like this - how to fix this?

**Reply by pbrzosko** [reply]
Seems like using coordGreedy and switching width with height produces more or less the layout that I want:
![image](https://github.com/erikbrinkman/d3-dag/assets/67122459/2564eebe-31aa-4573-94fe-5971d79c0438)

But do you know why the last node in the bottom subgraph is shifted?

**Reply by erikbrinkman** [reply]
1. For the tweaks, it'd be great to see the code associated with them. Your second image looks correct, just that the spacing is off. Potentially there's a bug with the way I'm handing spacing when applying the diagonal tweak, but it's hard to know without having something more concrete to work with.
2. the layout above is as expected. There is nothing to inherently center the graph, and so for for spits for two edges, it just picks one, this is inherent to the optimization.
3. `coordGreedy` is exactly what it sounds like, a greedy algorithm, so it's not perfect. The coordStep simply takes the layered dag and assigns y coordinates (in the flipped version). You're able to substitute your own if you don't like any of the outputs, but it's not trivial. You might also try `coordQuad` but I think it will make the tails bow outward, so it may not be preferable.

Also, I may have missed a question, so feel free to ask it again. This interface isn't great for asking several questions in a row, as I can't answer inline.
